R/teal_data-class.R

```diff
@@ -14,85 +14,61 @@ setOldClass("join_keys")
-#' @slot join_keys (`join_keys`) object specifying joining keys for data sets in `@env`.
+#' @slot join_keys (`join_keys`) object specifying joining keys for data sets in
+#' `@.xData`.
 #'  Access or modify with [join_keys()].
-#' @slot datanames (`character`) vector of names of data sets in `@env`.
-#'  Used internally to distinguish them from auxiliary variables.
-#'  Access or modify with [datanames()].
-#' @slot verified (`logical(1)`) flag signifying that code in `@code` has been proven to yield contents of `@env`.
+#' @slot verified (`logical(1)`) flag signifying that code in `@code` has been
+#'  proven to yield contents of `@.xData`.
 #'  Used internally. See [`verify()`] for more details.
 #'
 #' @import teal.code
 #' @keywords internal
 setClass(
   Class = "teal_data",
   contains = "qenv",
-  slots = c(join_keys = "join_keys", datanames = "character", verified = "logical"),
+  slots = c(join_keys = "join_keys", verified = "logical"),
   prototype = list(
     join_keys = join_keys(),
-    datanames = character(0),
     verified = logical(0)
   )
 )
 
-#' Initialize `teal_data` object
+#' It initializes the `teal_data` class
 #'
-#' @name new_teal_data
-#'
-#' @param data (`named list`) of data objects.
-#' @param code (`character` or `language`) code to reproduce the `data`.
-#'   Accepts and stores comments also.
-#' @param join_keys (`join_keys`) object
-#' @param datanames (`character`) names of datasets passed to `data`.
-#'   Needed when non-dataset objects are needed in the `env` slot.
-#' @rdname new_teal_data
-#' @keywords internal
-new_teal_data <- function(data,
-                          code = character(0),
-                          join_keys = join_keys(),
-                          datanames = names(data)) {
-  checkmate::assert_list(data)
-  checkmate::assert_class(join_keys, "join_keys")
-  if (is.null(datanames)) datanames <- character(0) # todo: allow to specify
-  checkmate::assert_character(datanames)
-  if (!any(is.language(code), is.character(code))) {
-    stop("`code` must be a character or language object.")
-  }
-
-  if (is.language(code)) {
-    code <- paste(lang2calls(code), collapse = "\n")
-  }
-  if (length(code)) {
-    code <- paste(code, collapse = "\n")
-  }
-  verified <- (length(code) == 0L && length(data) == 0L)
-
-  id <- sample.int(.Machine$integer.max, size = length(code))
+#' Accepts .xData as a list and converts it to an environment before initializing
+#' parent constructor (`qenv`).
+#' @noRd
+setMethod(
+  "initialize",
+  "teal_data",
+  function(.Object, .xData = list(), join_keys = join_keys(), ...) { # nolint: object_name.
+    # Allow .xData to be a list and convert it to an environment
+    if (!missing(.xData) && inherits(.xData, "list")) {
+      .xData <- rlang::env_clone(list2env(.xData), parent = parent.env(.GlobalEnv)) # nolint: object_name.
+      lockEnvironment(.xData, bindings = TRUE)
+    }
+    checkmate::assert_environment(.xData)
```

Comment:
I can't make a suggestion because GH doesn't get the differences properly. So I'm just attaching the code of the initialize. I'm just suggesting to utilise `callNextMethod` instead of adding `.Object@` afterwards 

```r
setMethod(
  "initialize",
  "teal_data",
  function(.Object, .xData = list(), join_keys = join_keys(), ...) { # nolint: object_name.
    # Allow .xData to be a list and convert it to an environment
    if (!missing(.xData) && inherits(.xData, "list")) {
      .xData <- rlang::env_clone(list2env(.xData), parent = parent.env(.GlobalEnv)) # nolint: object_name.
      lockEnvironment(.xData, bindings = TRUE)
    }
    args <- list(...)
    checkmate::assert_environment(.xData)
    checkmate::assert_class(join_keys, "join_keys")
    checkmate::assert_list(args, names = "named")
    methods::callNextMethod( # nolint: object_name.
      .Object,
      .xData,
      join_keys = join_keys,
      verified = (length(args$code) == 0L && length(.xData) == 0L),
      ...
    )
  }
)
```

Reply:
It makes sense, this way we delegate the actual slot assignments to the internals of S4 chain of `initialize()`

Done in 04fd26ef